firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationSquasher.java

```diff
@@ -0,0 +1,249 @@
+    setWithNoTransform.applyToLocalView(doc, Timestamp.now());
+
+    PatchMutation patchWithNoTransform =
+        new PatchMutation(
+            mutation.getKey(), mutation.getValue(), mutation.getMask(), mutation.getPrecondition());
+    patchWithNoTransform.applyToLocalView(doc, Timestamp.now());
+
+    // A Patch on a Set is still Set.
+    currentSquashed =
+        new SetMutation(
+            doc.getKey(),
+            doc.getData(),
+            Precondition.NONE,
+            mergeTransforms(
+                previouslySquashed.getFieldTransforms(), mutation.getFieldTransforms()));
+  }
+
+  private void squashPatchMutationWithPatch(
+      PatchMutation previouslySquashed, PatchMutation mutation) {
+    // Fields with transforms from `mutation` is taken out of the mask of `previouslySquashed` as
+    // they will get overwritten.
+    Set previousMask = previouslySquashed.getMask().getMask();
+    for (FieldTransform transform : mutation.getFieldTransforms()) {
+      if (previousMask.contains(transform.getFieldPath())) {
+        previousMask.remove(transform.getFieldPath());
+      }
+    }
+
+    // Fields in the mask from `mutation` is taken out of the transform list of `previouslySquashed`
+    // as they will get overwritten.
+    List<FieldTransform> previousTransforms = new ArrayList<>();
+    Set newMutationMask = mutation.getMask().getMask();
+    for (FieldTransform transform : previouslySquashed.getFieldTransforms()) {
+      if (!newMutationMask.contains(transform.getFieldPath())) {
+        previousTransforms.add(transform);
+      }
+    }
+
+    currentSquashed =
+        new PatchMutation(
+            mutation.getKey(),
+            mergeObjectValue(previouslySquashed.getValue(), mutation.getValue()),
+            mergeMask(FieldMask.fromSet(previousMask), mutation.getMask()),
+            previouslySquashed.getPrecondition(),
+            mergeTransforms(previousTransforms, mutation.getFieldTransforms()));
+  }
+
+  private static List<FieldTransform> mergeTransforms(
+      List<FieldTransform> previous, List<FieldTransform> newTransforms) {
+    Map<FieldPath, FieldTransform> transformMap = new HashMap<>();
+    for (FieldTransform t : previous) {
+      transformMap.put(t.getFieldPath(), t);
+    }
+
+    for (FieldTransform newTransform : newTransforms) {
+      transformMap.put(
+          newTransform.getFieldPath(),
+          mergeTransform(transformMap.get(newTransform.getFieldPath()), newTransform));
+    }
+
+    return new ArrayList(transformMap.values());
+  }
+
+  /** Merges a new transform into a previous transform on the same field. */
+  private static FieldTransform mergeTransform(
+      FieldTransform previous, FieldTransform newTransform) {
+    if (previous == null) {
+      return newTransform;
+    }
+
+    if (newTransform.getOperation() instanceof ServerTimestampOperation) {
+      return newTransform;
+    }
+
+    if (newTransform.getOperation() instanceof ArrayTransformOperation) {
```

Comment:
Do you think this logic can live in the individual transforms?

Reply:
Yeah.